### DISCUSSION TITLE
Are all icloud messages downloaded ?

### DISCUSSION THREAD

**Discussion by cedricloneux** [ReagentX/imessage-exporter#78] [Q&A] [answered]
Are all imessages downloaded (including those in icloud) ?
If not I would be useful to add an option to the CLI in order to download all messages locally first. Then works on this cache (the local db)
I don't think there is an optimal way to do it, such as APIs for downloading specific messages, for specific contacts etc... So we need to download the whole imessage data as cache first which will take disk space and won't be easy to clear later.

**Comment by ReagentX** [comment, marked as answer]
> Are all imessages downloaded (including those in icloud)?

No, the user must download them all. Apple does not expose any API for `imessage-exporter` to perform this step for you.

**Reply by cedricloneux** [reply]
do you know anyway to download all of it locally ? not sure simply disabling iCloud will do it

**Reply by ReagentX** [reply]
This option syncs messages for me: 

<img width="560" alt="image" src="https://user-images.githubusercontent.com/1920666/213580508-d4d06a78-7c04-4a91-b6f2-cc04d4af35b1.png">

**Reply by cedricloneux** [reply]
are really all conversations downloaded by default ? such as years before ? I highly doubt it.
I think only recent messages are downloaded and this button only manually update this lightweight cache.

**Reply by ReagentX** [reply]
When I enabled it on my new Mac it synced in everything from 2015 onwards, which is what I had synced to iCloud before.

**Comment by chuckbenz** [comment]
I've tried the tricks here (sync button; logging out and back in to iCloud in imessages on mac), but every time I run imessage-exporter on my mac, I only see messages back to Sept 12, 2022 in my family chat (the most active chat I have). On my iPad, I'm able to scroll back into Aug 2022 and earlier, so it's as if those messages are in the cloud, or at least on my iPad and should be syncing from there into the cloud? The iPad was set to keep messages for 1 year, but I've changed that to forever now... Anyone have further suggestions to get a deeper history on my mac? It's possible that Sept 12, 2022 is about when I setup that mac.

**Reply by chuckbenz** [reply]
I tried imessage-export on an older mac that I'd setup in Nov 2021, and it archived messages from then to sometime in early Dec 2021 (probably when I'd last had it turned on), and also more recent messages starting at Sept 12, 2022, but there is a 9 month gap in time where I don't have messages. I'm left wondering how to get the messages that the iPad/iPhone somehow have...
Still, it's good to get what I can, as a digital pack rat, I love this tool - thanks!

**Reply by chuckbenz** [reply]
I did find a time machine disk, and pulled messages from a chat.db on that disk, too. I still have an 8 month gap, but obviously I am well ahead of where I was.
